pkg/liqo-controller-manager/namespaceMap-controller/namespacemap_controller.go

```diff
@@ -18,69 +18,62 @@ package namespacemapctrl
 
-// NamespaceMapReconciler creates remote namespaces and updates NamespaceMaps.
+// NamespaceMapReconciler creates remote namespaces and updates NamespaceMaps and NamespaceOffloading Status.
 type NamespaceMapReconciler struct {
 	client.Client
 	Scheme        *runtime.Scheme
 	Mapper        meta.RESTMapper
 	RemoteClients map[string]client.Client
 }
 
-const (
-	clusterIDForeign = "discovery.liqo.io/cluster-id"
-)
-
 // Reconcile adds/removes NamespaceMap finalizer, and checks differences
-// between DesiredMapping and CurrentMapping.
+// between DesiredMapping and CurrentMapping in order to create/delete remote Namespaces if it is necessary.
 func (r *NamespaceMapReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	namespaceMap := &mapsv1alpha1.NamespaceMap{}
-	if err := r.Get(context.TODO(), req.NamespacedName, namespaceMap); err != nil {
+	if err := r.Get(ctx, req.NamespacedName, namespaceMap); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("The NamespaceMap '%s' doesn't exist anymore", req.Name)
+			return ctrl.Result{}, nil
+		}
 		klog.Errorf("%s --> Unable to get NamespaceMap '%s'", err, req.Name)
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{}, err
 	}
 
-	if !ctrlutils.ContainsFinalizer(namespaceMap, liqoconst.NamespaceMapControllerFinalizer) {
-		ctrlutils.AddFinalizer(namespaceMap, liqoconst.NamespaceMapControllerFinalizer)
-		if err := r.Patch(context.TODO(), namespaceMap, client.Merge); err != nil {
-			klog.Errorf("%s --> Unable to add finalizer to the NamespaceMap '%s'", err, namespaceMap.GetName())
-			return ctrl.Result{}, err
-		}
-		klog.Infof("Finalizer correctly added on NamespaceMap '%s'", namespaceMap.GetName())
+	// If someone deletes namespaceMap is necessary to remove all remote namespace
+	// associated with this resource before deleting it
+	if err := r.SetNamespaceMapControllerFinalizer(ctx, namespaceMap); err != nil {
+		return ctrl.Result{}, err
 	}
 
-	if err := r.manageRemoteNamespaces(namespaceMap); err != nil {
-		return ctrl.Result{}, err
+	// Create/Delete remote Namespaces if it is necessary, according to NamespaceMap status.
+	if r.ensureRemoteNamespacesExistence(namespaceMap) {
+		return ctrl.Result{}, fmt.Errorf("something during remote namespaces management went wrong")
 	}
 
-	if len(namespaceMap.Status.CurrentMapping) == 0 {
-		ctrlutils.RemoveFinalizer(namespaceMap, liqoconst.NamespaceMapControllerFinalizer)
-		if err := r.Update(context.TODO(), namespaceMap); err != nil {
-			klog.Errorf("%s --> Unable to remove '%s' from NamespaceMap '%s'", err,
-				liqoconst.NamespaceMapControllerFinalizer, namespaceMap.GetName())
-			return ctrl.Result{}, err
-		}
-		klog.Infof("Finalizer correctly removed from NamespaceMap '%s'", namespaceMap.GetName())
+	// Only when all remote namespaces associated with this NamespaceMap are deleted, remove finalizer
+	if err := r.RemoveNamespaceMapControllerFinalizer(ctx, namespaceMap); err != nil {
```

Comment:
IIUC, it is impossible to delete a NamespaceMap entry, unless all desired mapping are previously removed manually. Is this behavior desired? I doubt so

Reply:
It is impossible to delete a NamespaceMap, unless all remote Namespaces of that cluster are previously deleted, the finalizer doesn't care about DesiredMapping. So the finalizer is removed only when there are no more remote Namespace and the NamespaceMap is empty

Reply:
I personally feed this pretty strange, as it would be impossible to manually delete the resource. I would expect all mappings to be deleted upon termination, and then the finalizer to be removed. @palexster WDYT?